registry/registry.go

```diff
@@ -0,0 +1,125 @@
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/google/uuid"
+	"github.com/tcsc/levity/task"
+)
+
+// handleGeneratorFn represents an arbitrary function that can be called to
+// generate a candidate TaskHandle value.
+type handleGeneratorFn func() string
+
+// Registry models a simple task registration system. For a simple exercise
+// like this we are just keeping the task information in memory. This would
+// obviously require somthing more durable in a production server.
+type Registry struct {
+	lock           sync.RWMutex
+	db             map[string]*task.Task
+	generateHandle handleGeneratorFn
+}
+
+func handleFromUUID() string {
+	return uuid.New().String()
+}
+
+// New creates and returns an initialised, ready-to-use task registry
+func New() *Registry {
+	return newWithIDGenerator(handleFromUUID)
+}
+
+func newWithIDGenerator(gen handleGeneratorFn) *Registry {
+	return &Registry{
+		db:             make(map[string]*task.Task),
+		generateHandle: gen,
+	}
+}
+
+// Lookup attempts to find a task from its ID. Returns nil if no such task exists.
+func (registry *Registry) Lookup(handle string) *task.Task {
+	registry.lock.RLock()
+	defer registry.lock.RUnlock()
+
+	if t, exists := registry.db[handle]; exists {
+		return t
+	}
+
+	return nil
+}
+
+func isDone(ctx context.Context) bool {
+	select {
+	case <-ctx.Done():
+		return true
+
+	default:
+		return false
+	}
+}
+
+// AllocateID attempts to allocate a unique task ID in the registry. Keeps
+// trying to generate a unique ID until it succeeds, or the supplied context
+// expires.
+func (registry *Registry) AllocateID(ctx context.Context) (string, error) {
+	// For the purposes of this exercise we're using a stringified UUID as
+	// the task ID. This should be statistically unique - certainly we
+	// should not expect to see a collision in the lifetime of the server.
+	//
+	// To handle the remote possibility of a collision, we keep trying until
+	// we have a unique ID registered in the database. To handle the even more
+	// remote possibility of repeated collisions, we will only keep trying
+	// until the supplied context expires
+	for !isDone(ctx) {
+		candidate := registry.generateHandle()
+		if registry.tryAllocateID(candidate) {
+			return candidate, nil
+		}
+	}
+	return "", ctx.Err()
+}
+
+// tryRegister makes a single attempt to register a task with the supplied ID.
+func (registry *Registry) tryAllocateID(id string) bool {
+	registry.lock.Lock()
+	defer registry.lock.Unlock()
+
+	if _, exists := registry.db[id]; exists {
+		return false
+	}
+
+	registry.db[id] = nil
+	return true
+}
+
+var UnallocatedIDError = errors.New("ID not allocated")
+var AlreadyBoundIDError = errors.New("ID already bound to a task")
+
+// BindToID binds a task to a pre-allocated ID. It is considered an error to
+// attempt to bind a tsk to an ID that has not been pre-allocated, or to an
+// ID already bound to a task.
+func (registry *Registry) BindToID(id string, t *task.Task) error {
```

Comment:
Whats the usecase for having IDs be separately allocated in advance?  Seems like this one feature adds a lot of complexity.

Reply:
It does. Most of it stems from me treating the task IDs as an arbitrary string, rather than a stringified UUID.

If you think of the ID as an arbitrary string, then the ID generation could potentially fail (e.g. if you had _really_ bad luck and continually generated clashing ID strings until the context timed out). Anyway, I didn't want to have to start the task process and then immediately kill it if the registration failed, so I split the registration into two parts; allocating the ID in advance (that could fail) and binding the task to the ID once it was started (which could not fail, at least if used correctly).

But yeah, it's less than ideal.

Following @russjones' suggestion to just trust that the generated UUIDs are actually unique drastically simplifies the whole thing and pretty much all of this is now deleted.